practice_app/api/views.py

```diff
@@ -1,3 +1,109 @@
+        response = {}
+        response["user_id"] = user_id
+        response["name"] = api_request["person"]["name"]["given-names"]["value"]
+
+        if api_request["person"]["name"]["family-name"] != None:
+            response["surname"] = api_request["person"]["name"]["family-name"]["value"]
+        else:
+            response["surname"] = None
+        return JsonResponse(response)
+
+# POST api/log_in/
+# implements log in functionality
+# username and password should be provided in Headers
+def log_in(request):
+    
+    username = request.headers["username"]
+    password = request.headers["password"]
+
+    if username == None or username == '':
+        return JsonResponse({"status":"ORCID ID should be provided."}, status = 404)
+    
+    if password == None or password == '':
+        return JsonResponse({"status":"Password should be provided."}, status = 404)
+    
+    # Check user database to authenticate the user with given credentials. Return a user if valid username and password is given.
+    user = authenticate(request, username=username, password=password)
+
+    # if the user is authenticated log in by built-in login function
+    if user is not None:
+        login(request, user)
+        return JsonResponse({"status":"User logged in."}, status = 200)
+
+    else:
+        return JsonResponse({"status":"Authentication failed"}, status = 404)
+
+# GET api/log_out/
+# get user to log out by built-in logout function
+def log_out(request):
+
+    logout(request)
+    return JsonResponse({"status":"User logged out."}, status = 200)
+
+# POST api/user_registration
+# username and password should be in headers
+# username should be unique
+# name and surname should be provided in data
+# username, name, password must be provided, surname is optional
+def user_registration(request):
+
+    user_id = request.headers["username"]
+    password = request.headers['password']
+    name = request.POST['name']
+    surname = request.POST['surname']
+
+    if user_id == None or user_id == '':
+        return JsonResponse({"status":"ORCID ID should be provided."}, status = 404)
+    
+    if password == None or password == '':
+        return JsonResponse({"status":"Password should be provided."}, status = 404)
+    
+    if name == None or name == '':
+        return JsonResponse({"status":"Name should be provided."}, status = 404)
+    
+    if surname == None:
+        surname = " "
+    
+    # Check if the user_id is already taken. 
+    if len(User.objects.filter(username= user_id)) == 0:
+        # Create user on User model by using create_user function.
+        # Used create_user() instead of create() since create_user() handles password encryption
+        User.objects.create_user(username= user_id, password= password, first_name= name,last_name= surname )
+        return JsonResponse({"status":"User created"}, status = 200)
+    
+    else:
+        return JsonResponse({"status":"Username is already taken."}, status = 404)
```

Comment:
I think the status code shouldn't be 404 here since a username being already taken isn't really a not found error.
This may be helpful instead: https://stackoverflow.com/questions/12657493/what-http-error-code-to-return-for-name-already-taken